### ISSUE TITLE
Background images in confusion matrix

### ISSUE THREAD

**Issue by robertastellino** [ultralytics/ultralytics#4960]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have a question about the background images in the confusion matrix. 
I put in my dataset (both in training and validation set) some background images that are not labeled because there is no object to detect. I tried both including empty labels `.txt `(background images have the corrisponding - empty - label in the labels folder) and not including them at all (so in this case the background images don't have the corrisponding label `.txt` in the labels folder). In both cases, when I run the CLI command to train, the number of background images are taken into account correctly: 

`train: Scanning ./labels/train.cache... 6170 images, 88 backgrounds, 0 corrupt: 100% 6170/6170 [00:00<?, ?it/s] `
`val: Scanning ./labels/val.cache... 772 images, 12 backgrounds, 0 corrupt: 100% 772/772 [00:00<?, ?it/s]`

![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/138810693/ce63b1a6-5f4e-4c24-939d-9687e6816379)

I wonder how does YOLOv8 treat the backgroung images? And why does it misclassify all the background images (the green part is always zero)? How should I interpret this confusion matrix?

**Comment by glenn-jocher** [comment]
Hi @robertastellino,

Thank you for trying out the suggestions and for your follow-up question.

In terms of the number of backgrounds (images without any objects of interest), it's not necessarily about having an equal number of backgrounds and non-backgrounds. The ratio between your background and non-background images depends on the nature of the task and the specifics of your dataset.

However, it is crucial to have a representative distribution that matches the real-world incidence of your classes. As a simple rule, if empty images (backgrounds) are a common occurrence in the environments where you plan to use the model, it would be beneficial to include a significant amount of them in your dataset.

Following this logic, if your model will often deal with background images in its predicted use case, having 88 training and 12 validation background images out of 6170 and 772 total images respectively should not be a problem, and could possibly be quite representative of the deployment scenario.

That being said, your results might improve by increasing the number of examples of classes you want to detect, as you mentioned. Experimenting with the number of examples per class, including backgrounds, could help adjust your model to work better.

I hope this helps! If you have any other questions, feel free to ask.

Best regards,

Glenn Jocher

**Comment by mmacander** [comment]
Glenn, you say that "The 'green part being always zero' suggests that YOLOv8 never correctly classifies an image as a background when it indeed is a background. It is also possible that the model might be interpreting these images as containing objects from existing classes, causing the 'misclassification'. " And you go on to suggest ways to improve the model.

However, I've never see a yolov8 confusion matrix in my projects or in these forums where the green part was anything other than zero/white. No matter how many background images are included in a validation run, it never includes any images that were correctly determined to have zero objects in the confusion matrix. Am I missing something here?

**Comment by glenn-jocher** [comment]
@mmacander hi there,

Thanks for reaching out with your observation! It's indeed curious that the confusion matrix consistently shows zero for correctly classified background images across different projects. This could be an indication that the model is always predicting at least one object, even in background images, which would mean it's never predicting 'no objects' correctly.

One possible step to investigate this further is to check the predictions directly. You could run the model on your background images and see if it's always predicting objects when it shouldn't. Here's a quick way to do this for a single image:

```bash
yolo detect predict model=yolov8n.pt source=path/to/your/background.jpg
```

If the model is indeed predicting objects in every background image, it might be worth revisiting the balance of your dataset or considering a more thorough hyperparameter tuning.

Let me know if this helps or if you have any more questions!

**Comment by mmacander** [comment]
The model is correctly predicting most (not all) background images as containing no objects. The confusion matrix does not reflect that. The green box is always empty. If this was a GPT answer it's an example of a very misleading and confusing one.

**Comment by mmacander** [comment]
Raw output and confusion matrix from a validation run are below. The distribution of images in this run matches the distribution in the full dataset (about 90% of images in my domain are empty). For training the model, the ratio was 80% images with objects and 20% empty images. There are empty .txt files matching all the empty images.

I also did a val run with a distribution that matched the training data (80% with objects/20% empty). For the run below, I added a couple thousand empty images and kept the exact same set of images with objects. The confusion matrix for both runs was identical (background/background = 0 or white). The F1-confidence and related plots are quite a bit different though, so the empty images are definitely included in part of the validation pipeline.

$ yolo detect model=${model_path}   mode=val   split=test   project=results/val   name=test_90pct_empty_ioup4   data=dataset.yaml   exist_ok=True   imgsz=640   plots=True   batch=32    iou=0.4   save_json=True   save_txt=True


Ultralytics YOLOv8.1.14 🚀 Python-3.11.6 torch-2.2.0+cu121 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)
Model summary (fused): 268 layers, 43608150 parameters, 0 gradients, 164.8 GFLOPs
val: Scanning labels/test... 2300 images, 2069 backgrounds, 0 corrupt: 100%|██████████| 2300/2300 [00:02<00:00
val: New cache created: labels/test.cache
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 72/72 [05:00<00:00,  4.18s/it]
                   all       2300        933        0.8      0.854      0.877      0.615
            ground_bwb       2300        881      0.802      0.843      0.871      0.597
            flying_bwb       2300         52      0.798      0.865      0.883      0.634
Speed: 0.1ms preprocess, 4.9ms inference, 0.0ms loss, 0.2ms postprocess per image
Saving results/val/test_90pct_empty_ioup4/predictions.json...
Results saved to results/val/test_90pct_empty_ioup4
💡 Learn more at https://docs.ultralytics.com/modes/val


![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/25831875/12e57d0d-4ce3-4f8d-9877-b3b7bfda40c2)

**Comment by glenn-jocher** [comment]
@mmacander hi there,

Thanks for sharing the details of your validation runs and the confusion matrix behavior. It's interesting that the F1-confidence plots differ while the confusion matrix remains unchanged, suggesting that the empty images are indeed influencing some aspects of the validation.

Since the confusion matrix isn't reflecting the correct classification of background images, this might be an issue with how the matrix is computed or visualized. It's great that you've checked the raw output and confirmed that the empty images are being processed.

I recommend opening an issue on the Ultralytics YOLOv8 GitHub repository with the details you've provided here. The team can then look into why the confusion matrix isn't updating as expected. In the meantime, you can rely on the raw output and other plots for a more accurate assessment of your model's performance.

Keep up the great work, and we appreciate your contributions to improving YOLOv8! 🌟

**Comment by YanzhenHuang** [comment]
@glenn-jocher 

Dear author:

I have the exact same question of the **Missing `background` tile in confusion matrix**. 

This is my confusion matrix:

<img src="https://github.com/user-attachments/assets/f84bb6c4-c6a1-4fdd-b38e-081e5fa0e226" width="500px">

This is how my  dataset is split:

![Image](https://github.com/user-attachments/assets/1b00992b-0763-4d31-9e88-51e222defc6a)


**I have the following information:**
1. I use Roboflow to annotate my dataset to train the model to specifically recognize cell phones. And `cell-phones` is the only positive class.
2. The negative data are all marked as NULL in Roboflow (please see the image below), which results in an empty `.txt` file after downloading. These data are images containing random objects that I am indicating the model not to pay any attention to (such as bottles, cans, etc.). 
3. The model performs fairly well in application.

This is how the negative data is labeled: A NULL class, with absolutely no bounding boxes.
<img src="https://github.com/user-attachments/assets/42b69cba-1f1c-4cad-8106-d6f756a0aff2" width="500px">


**I want to ask:**
1. Is the absence of the `background` tile an indication of some errors, or is it designed this way?
2. If it is designed this way, is the mode **is designed** not to classify any `background`?

**Comment by glenn-jocher** [comment]
Hi @YanzhenHuang,

Thank you for providing those detailed screenshots and information about your dataset setup. What you're observing with the missing background tile in the confusion matrix is actually expected behavior in the YOLO detection task.

In the detection framework, the confusion matrix is designed to track relationships between predicted objects and ground truth objects. The background (images with no objects or empty .txt files) is handled differently than regular classes. When the model correctly identifies that there's nothing to detect in a background image, this isn't explicitly represented in the confusion matrix visualization.

Looking at the code in `ultralytics/utils/metrics.py`, the background class is included in the matrix dimensions (`nc+1, nc+1` for detection tasks), but the visualization doesn't show true negative cases (correctly identified backgrounds) in the typical way you might expect from classification tasks.

This doesn't indicate an error in your implementation - your model is still evaluating those background images during validation, and they're affecting metrics like precision and recall. The confusion matrix simply doesn't highlight background true negatives visually.

Since your model is performing well in application, I wouldn't be concerned about this visualization limitation.
